### PR TITLE
stdenv: Avoid some list allocations in check-meta when checking licenses

### PR DESCRIPTION
**Pull request by adisbladis**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/stdenv/generic/check-meta.nix [hunk 1]
```diff
@@ -34,23 +36,29 @@ let
...
   allowBroken = config.allowBroken
     || builtins.getEnv "NIXPKGS_ALLOW_BROKEN" == "1";
 
   allowUnsupportedSystem = config.allowUnsupportedSystem
     || builtins.getEnv "NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM" == "1";
 
-  isUnfree = licenses: lib.lists.any (l: !l.free or true) licenses;
+  isUnfree = licenses:
+    if isAttrs licenses then !licenses.free or true
+    # TODO: Returning false in the case of a string is a bug that should be fixed.
```
Comment (adisbladis) [inline pkgs/stdenv/generic/check-meta.nix:54]: This bug has been explored/documented at https://gist.github.com/adisbladis/b5aa64ded97712e062728bd033e90fa2.
I will create a more proper tracking issue for it shortly.

Comment (adisbladis) [reply pkgs/stdenv/generic/check-meta.nix:54]: https://github.com/NixOS/nixpkgs/issues/269788

File: pkgs/stdenv/generic/check-meta.nix [hunk 2]
```diff
@@ -34,23 +36,29 @@ let
   hasLicense = attrs:
     attrs ? meta.license;
 
-  hasAllowlistedLicense = assert areLicenseListsValid; attrs:
-    hasLicense attrs && lib.lists.any (l: builtins.elem l allowlist) (lib.lists.toList attrs.meta.license);
+  hasListedLicense = assert areLicenseListsValid; list: attrs:
+    length list > 0 && hasLicense attrs && any (l: elem l list) (toList attrs.meta.license);
```
Comment (piegamesde) [inline pkgs/stdenv/generic/check-meta.nix:40]: Is there a non-ugly way to get rid of that `toList` as well?

Comment (adisbladis) [reply pkgs/stdenv/generic/check-meta.nix]: Done. Though it's worth noting that this branch is only hit when you have either allow/block listed licenses, so I'd wager the impact is tiny.

**Comment by piegamesde** [comment]
In a quest to find more systemic ways to do these optimizations: could singleton lists be special cased in the interpreter to improve performance?

**Comment by adisbladis** [comment]
> could singleton lists be special cased in the interpreter to improve performance?

They are already.

**Comment by piegamesde** [comment]
Hm, okay. Alternatively, could the singleton handling logic be pushed down into the lib functions like `any` and `all`? As in, to make them work with a single item instead of a list, so that all `toList` calls on the call site may be omitted, but also without having to manually unwrap the logic at call site. (I do think this change makes the code worse, but I'd accept it for the performance benefits if there is no better solution)

**Pull request merged**